ygot/diff_test.go

```diff
@@ -0,0 +1,263 @@
+	MapValue    map[string]*basicListMember `path:"map-list"`
+}
+
+func (*basicStruct) IsYANGGoStruct() {}
+
+type basicStructTwo struct {
+	StringValue *string           `path:"second-string-value"`
+	StructValue *basicStructThree `path:"struct-three-value"`
+}
+
+type basicListMember struct {
+	ListKey *string `path:"list-key"`
+}
+
+func (*basicListMember) IsYANGGoStruct() {}
+func (b *basicListMember) ΛListKeyMap() (map[string]interface{}, error) {
+	return map[string]interface{}{
+		"list-key": *b.ListKey,
+	}, nil
+}
+
+type errorListMember struct {
+	StringValue *string `path:"error-list-key"`
+}
+
+func (*errorListMember) IsYANGGoStruct() {}
+func (b *errorListMember) ΛListKeyMap() (map[string]interface{}, error) {
+	return nil, fmt.Errorf("invalid key map")
+}
+
+type badListKeyType struct {
+	Value *complex128 `path:"error-list-key"`
+}
+
+func (*badListKeyType) IsYANGGoStruct() {}
+func (b *badListKeyType) ΛListKeyMap() (map[string]interface{}, error) {
+	return map[string]interface{}{
+		"error-list-key": *b.Value,
+	}, nil
+}
+
+type basicStructThree struct {
+	StringValue *string `path:"third-string-value|config/third-string-value"`
+}
+
+func TestNodeValuePath(t *testing.T) {
+	tests := []struct {
+		desc          string
+		inNI          *util.NodeInfo
+		inSchemaPaths [][]string
+		wantPathSpec  *pathSpec
+		wantErr       string
+	}{{
+		desc: "root level element",
+		inNI: &util.NodeInfo{
+			Parent: nil,
+		},
+		inSchemaPaths: [][]string{{"one", "two"}, {"three", "four"}},
+		wantPathSpec: &pathSpec{
+			gNMIPaths: []*gnmipb.Path{{
+				Elem: []*gnmipb.PathElem{{Name: "one"}, {Name: "two"}},
+			}, {
+				Elem: []*gnmipb.PathElem{{Name: "three"}, {Name: "four"}},
+			}},
+		},
+	}, {
+		desc: "nodeinfo missing parent annotation",
+		inNI: &util.NodeInfo{
+			Parent: &util.NodeInfo{
+				Annotation: []interface{}{},
+			},
+		},
+		wantErr: "could not find path specification annotation",
+	}, {
+		desc: "nodeinfo for a child path",
+		inNI: &util.NodeInfo{
+			Parent: &util.NodeInfo{
+				Annotation: []interface{}{
+					&pathSpec{
+						gNMIPaths: []*gnmipb.Path{{
+							Elem: []*gnmipb.PathElem{{
+								Name: "parent",
+							}},
+						}},
+					},
+				},
+			},
+		},
+		inSchemaPaths: [][]string{{"foo", "bar"}, {"baz"}},
+		wantPathSpec: &pathSpec{
+			gNMIPaths: []*gnmipb.Path{{
+				Elem: []*gnmipb.PathElem{{Name: "parent"}, {Name: "foo"}, {Name: "bar"}},
+			}, {
+				Elem: []*gnmipb.PathElem{{Name: "parent"}, {Name: "baz"}},
+			}},
+		},
+	}, {
+		desc: "nodeinfo for a child path missing annotation path",
+		inNI: &util.NodeInfo{
+			Parent: &util.NodeInfo{
+				Annotation: []interface{}{},
+			},
+		},
+		inSchemaPaths: [][]string{{"foo", "bar"}, {"baz"}},
+		wantErr:       "could not find path specification annotation",
+	}}
+
+	for _, tt := range tests {
+		got, err := nodeValuePath(tt.inNI, tt.inSchemaPaths)
+		if err != nil && err.Error() != tt.wantErr {
```

Comment:
(optional) Should we prefer to test error semantics rather than the error value? Testing the error value can turn the test into a change detector where with a spelling change in the error message the author also needs to update tests.

Reply:
This is the pattern that we've used elsewhere. The semantics are difficult to test -- since in general we're just returning an `error` type. We want to make sure that the error that we catch is the one that exercises the specific error handling. How do you suggest we check that without checking the contents without defining new types for every error handling branch (which seems maintenance overhead)?